src/cargo/ops/cargo_package.rs

```diff
@@ -816,7 +816,8 @@ fn check_repo_state(
         // - ignored (in case the user has an `include` directive that
         //   conflicts with .gitignore).
         let mut dirty_files = Vec::new();
-        collect_statuses(repo, &mut dirty_files)?;
+        let pathspec = relative_pathspec(repo, pkg.root());
+        collect_statuses(repo, &[pathspec.as_str()], &mut dirty_files)?;
```

Comment:
What if the readme is `../README.md`?

Reply:
I don't think this PR changed anything for relative paths. It is a bug itself.

They will be filtered out later in

https://github.com/rust-lang/cargo/blob/59b2ddd4484a04a0c202052202cc9f30ac48165c/src/cargo/ops/cargo_package.rs#L829

That said, I don't think either dirty symlinks or relative paths were correctly handled on master.

Reply:
Let me create a test an verify it.

Reply:
That is saying if a `src_file` should start with a `dirty_file` path, else filter it out.  I'm not seeing how thats relevant to a `../README.md`.  It would if `Cargo.toml` is in the repo root but it is most likely to be in a sub-directory within the repo, so that `README.md` could be among `dirty_files`.

Or is `src_files` not including these `../README.md` files?

Reply:
Either way, isn't this cementing the bug in further, making it harder for us to fix?

Reply:
`src_files` includes symlink files as ordinary files. By the time we get `src_files`, file metadata has been lost.

I don't think this change worsen the situation, unless we consider rewriting `list_files` to do more as a fix to #14955.

Reply:
What about the non-symlink case?

Reply:
See #14967.

Going to close this and discuss the performance issue in #14955